numpy/core/src/multiarray/compiled_base.c

```diff
@@ -1103,23 +1103,37 @@ arr_unravel_index(PyObject *self, PyObject *args, PyObject *kwds)
 
     unravel_size = PyArray_MultiplyList(dimensions.ptr, dimensions.len);
 
+    dtype = PyArray_DescrFromType(NPY_INTP);
+    if (dtype == NULL) {
+        goto fail;
+    }
+
     if (!PyArray_Check(indices0)) {
         indices = (PyArrayObject*)PyArray_FromAny(indices0,
                                                     NULL, 0, 0, 0, NULL);
         if (indices == NULL) {
             goto fail;
         }
+
+        if (PyArray_SIZE(indices) == 0) {
+            // Unfortunately, if the input array is empty, then indices would
+            // have a default dtype of float64, which isn't compatibile.
+            // Force a recast of an empty array.
+            PyArrayObject *indices_tmp = NULL;
+            indices_tmp = (PyArrayObject*)PyArray_FromArray(indices, dtype,
```

Comment:
I think you probably have to incref dtype here. IIRC FromArray steals a reference. It seems good otherwise, this code solves the same issue:
https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/mapping.c#L376